projects/hathor-wallet-mobile/0002-nano-contract-integration.md

```diff
@@ -0,0 +1,476 @@
+	 	- *Transaction ID*
+		- *Nano Contract ID*
+		- *Blueprint Method*
+		- *Argument List*
+		- *Timestamp*
+
+## Register Nano Contract
+
+##### When execute
+- It should happen when a *Nano Contract* is registered using the *Nano Contract Registration* flow.
+
+##### Where to execute
+- On `sagas/nanoContract.js@registerNanoContract`.
+
+##### Where to store
+- On redux state `registeredNanoContracts`.
+- On AsyncStore using a key `nanoContract:registered` for persistence, once the lib doesn't have an interface on the `store`.
+
+## Wallet Connect
+
+### RPC Protocol
+
+#### `htr_createNcTx`
+[htr_createnctx]: #htr_createnctx
+
+The `htr_createNcTx` method creates a new *Nano Contract* transaction based on parameters. It will select the utxos automatically, mine and send the transaction.
+
+##### Parameters
+
+- `blueprintId`: unique identification of the *Blueprint*
+- `method`: *Blueprint*'s method name
+- `caller`: wallet address representing method's *Caller*
+- `ncId`: unique identification of *Nano Contract*, it is not required when method is equal `initialize`
+- `ncActions`: a list of *Actions* the *Nano Contract* must process
+	- `type`: type of action, either `deposit` or `withdrawal`
+	- `token`: unique identification of the *Token*
+	- `data`: detailed information for the action
+		- `amount`: token's amount
+		- `address`: wallet's address, it is required only for `withdrawal` and ignored for `deposit`
+- `ncArgs`: a positional list of method's *Arguments*
+	- `type`: a supported type, like `boolean`, `number`, `binary`, etc.
+	- `value`: type's literal value
+
+##### Example Request
+
+```json
+{
+  "id": 3,
+  "jsonrpc": "2.0",
+  "method": "htr_createNcTx",
+  "params": {
+	"bluprintId": "494d0ac59a6918b771b122395206fef3f349f84f20dc430188a319d4ead24a3b",
+	"method": "swap",
+	"caller": "WRqBSi5jc74P8tRo7CmgiGByA1eTzfMh7B",
+	"ncId": "000001e43dda7348c53c6252d98799c9e2a0b4ea3b5f10b1182578d643c9641d",
+	"ncActions": [
+		{
+			"type": "deposit",
+			"token": "000001a176aa72b84f3486b3fcea327b023a70d1b6461c1b6a9073785b5775d8",
+			"data": {
+				"amount": "100",
+			},
+		},
+		{
+			"type": "withdrawal",
+			"token": "000001a176aa72b84f3486b3fcea327b023a70d1b6461c1b6a9073785b5775d8",
+			"data": {
+				"amount": "200",
+				"address": "WRqBSi5jc74P8tRo7CmgiGByA1eTzfMh7B",
+			},
+		},
+	],
+	"ncArgs": [],
+}
+```
+
+### Nano Contract Transaction Request
+
+##### When execute
+- Every time the a connected dapp send the request `htr_createNcTx`.
+
+##### Where to execute
+- It must be handled on `sagas/walletConnect.js@onSessionRequest` and dispatched to `onCreateNcTxRequest`.
+- It must be processed on `sagas/walletConnect.js@onCreateNcTxRequest`.
+
+##### How to execute `onCreateNcTxRequest`
+1. Validate the transaction request
+2. Ask user for `Approval` or `Rejection`
+3. If rejected, return to dapp an error message
+4. If approved, ask user's Pin Code
+5. Create and send transaction
+6. Return to dapp a success message 
```

Comment:
In this specific example, shouldn't the user approve the use of his utxos?

Reply:
Oh! Yes! I didn't stressed this part because after the request be approved almost immediately the PIN will be required, but I will make it clear.

Reply:
Humm... I see this operation like the send transaction. The use didn't see the UTXOs, the wallet chose the UTXOs to be used. I mimic this behavior there:

```
4. If approved, ask user's Pin Code
```

Reply:
Do you believe this is necessary? For this we would need a method on the wallet lib side just to mount the transaction and another just to sign and publish, other than we would need more screens in this project to carry on this behavior.

Reply:
No, you're right, we don't ask for confirmation on the UTXOs the lib selected in the send transaction method, so we shouldn't here either